File: 077/cpp/main.cpp
class Solution {
public:
    vector<vector<int>> combine(vector<int> ns, int k) {
        if (ns.size() < k) {
            return {};
        }
        if (ns.size() == k) {
            return {ns};
        }
        int n = ns.back();
        ns.pop_back();
        auto result = combine(ns, k);
        auto result0 = combine(ns, k - 1);
        for (auto &r0: result0) {
            r0.push_back(n);
            result.push_back(r0);
        }
        return result;
    }

    vector<vector<int>> combine(int n, int k) {
        vector<int> ns;
        for (int i = 0; i < n; i++) {
            ns.push_back(i + 1);
        }
        return combine(ns, k);
    }
};

File: 077/rust/main.rs
impl Solution {

    pub fn combine2(ns: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        if ns.len() < k as usize {
            return Vec::new();
        }
        if ns.len() == k as usize {
            return vec![ns];
        }
        let mut ns = ns;
        let n = ns[0];
        ns.remove(0);
        let mut result = Solution::combine2(ns.clone(), k);
        let result0 = Solution::combine2(ns.clone(), k - 1);
        for r0 in result0 {
            let mut r = r0.clone();
            r.push(n);
            result.push(r);
        } 
        return result;
    }

    pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
        let ns: Vec<i32> = (1..n+1).collect();
        return Solution::combine2(ns, k);
    }
}
